packages/backend/src/modules/finality/analyzers/polygon-zkevm/PolygonZkevmFinalityAnalyzer.ts

```diff
@@ -0,0 +1,112 @@
+import {
+  notUndefined,
+  ProjectId,
+  TrackedTxsConfigSubtype,
+  UnixTime,
+} from '@l2beat/shared-pure'
+import { utils } from 'ethers'
+import { z } from 'zod'
+
+import { RpcClient } from '../../../../peripherals/rpcclient/RpcClient'
+import { LivenessRepository } from '../../../tracked-txs/modules/liveness/repositories/LivenessRepository'
+import { byteArrFromHexStr } from '../opStack/utils'
+import { BaseAnalyzer } from '../types/BaseAnalyzer'
+import { decodeBatch } from './batch'
+import { toTransactionHash } from './hash'
+
+export class PolygonZkEvmFinalityAnalyzer extends BaseAnalyzer {
+  constructor(
+    provider: RpcClient,
+    livenessRepository: LivenessRepository,
+    projectId: ProjectId,
+    private readonly l2Provider: RpcClient,
+  ) {
+    super(provider, livenessRepository, projectId)
+  }
+
+  override getTrackedTxSubtype(): TrackedTxsConfigSubtype {
+    return 'batchSubmissions'
+  }
+
+  async getFinality(transaction: {
+    txHash: string
+    timestamp: UnixTime
+  }): Promise<number[]> {
+    const tx = await this.provider.getTransaction(transaction.txHash)
+    const l1Timestamp = transaction.timestamp
+
+    const hashes = extractTransactionData(tx.data)
+      .map(byteArrFromHexStr)
+      // Might be memory intensive - if so, consider batching/sequential processing
+      .flatMap(decodeBatch)
+      .flatMap((block) => block.transactions)
+      .map(toTransactionHash)
+
+    //TODO: poor stuff given the batch can contain ~250 txs, either make sure its rate limited or pick at random a few txs
```

Comment:
Aren't the transactions sorted? Then it's enough to know the first and last to know the min and max. It should be also enough for avg. This call looks a lot too big.

Reply:
Especially that we only download the max block. Please check how are the txs usually sorted.

Reply:
min and max are enough for the avg only if the distribution is uniform and i'm not sure this is an assumption that approximates good enough